src/shared/dissect-image.c

```diff
@@ -841,84 +845,110 @@ static int dissect_image(
 
         uint64_t diskseq = m->loop ? m->loop->diskseq : 0;
 
-        if (verity && iovec_is_set(&verity->root_hash)) {
-                sd_id128_t fsuuid, vuuid;
+        /* If input is a single already open verity device and the filesystem type is pinned to a single
+         * fstype, and it's a single-image filesystem, then skip all the expensive probing and just shortcut
+         * to mount the dm device directly */
+        if ((flags & DISSECT_IMAGE_NO_PARTITION_TABLE)) {
+                PartitionPolicyFlags policy_flags = image_policy_get_exhaustively(policy, PARTITION_ROOT);
+                if (policy_flags < 0)
+                        return policy_flags;
+
+                /* The policy fstype flags translate to the literal fstype name of each filesystem. */
+                r = partition_policy_flags_to_string(policy_flags & _PARTITION_POLICY_FSTYPE_MASK, /* simplify= */ true, &root_fstype_string);
+                if (r < 0)
+                        return r;
+                /* If no image FD was passed in, then the input device must be a dm-verity node
+                 * with the other artifacts and a single filesystem */
+                if (fd < 0 && !(verity && verity->data_path &&
+                                r == 1 && /* A single fstype was specified */
+                                path_startswith(devname, "/dev/mapper/")))
+                        return -ENOPKG;
+                /* If there are multiple filesystems in the policy then don't set these and go to blkid */
+                if (r == 1) {
+                        root_fs_usage = policy_flags & PARTITION_POLICY_ENCRYPTED ? "crypto" : "filesystem";
+                        root_fstype = policy_flags & _PARTITION_POLICY_FSTYPE_MASK;
+                }
+        }
+
+        if (!root_fstype) {
```

Comment:
should be `if (!root_fstyle || !root_fs_usage)` no?

Reply:
if root_fstype is set then fs_usage is guaranteed to be set, but I can change it, doesn't hurt to be precise